Documentation for the OAuth2 integration (#1113)
api-platform/docs: core/oauth2.md

     * @ORM\ManyToOne(targetEntity="AppBundle\Entity\User")
     */
    protected $user;
}
```

```php
<?php
// AppBundle/Entity/AuthCode.php

namespace AppBundle\Entity;

use FOS\OAuthServerBundle\Entity\AuthCode as BaseAuthCode;
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity
 */
class AuthCode extends BaseAuthCode
{
    /**
     * @ORM\Id
     * @ORM\Column(type="integer")
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    protected $id;

    /**
     * @ORM\ManyToOne(targetEntity="Client")
     * @ORM\JoinColumn(nullable=false)
     */
    protected $client;

    /**
     * @ORM\ManyToOne(targetEntity="AppBundle\Entity\User")
     */
    protected $user;
}

```

```php
<?php
// AppBundle/Entity/Client.php

namespace AppBundle\Entity;

use FOS\OAuthServerBundle\Entity\Client as BaseClient;
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity
 */
class Client extends BaseClient
{
    /**
     * @ORM\Id
     * @ORM\Column(type="integer")
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    protected $id;

    public function __construct()
    {
        parent::__construct();
        // your own logic
    }
}

```

```php
<?php
// AppBundle/Entity/RefreshToken.php

> namespace AppBundle\Entity;

Reviewer: ```suggestion
namespace App\Entity;
```

Same everywhere

use FOS\OAuthServerBundle\Entity\RefreshToken as BaseRefreshToken;
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity
 */
class RefreshToken extends BaseRefreshToken
{
    /**
     * @ORM\Id
     * @ORM\Column(type="integer")
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    protected $id;

    /**
     * @ORM\ManyToOne(targetEntity="Client")
     * @ORM\JoinColumn(nullable=false)
     */
    protected $client;

    /**
     * @ORM\ManyToOne(targetEntity="AppBundle\Entity\User")
     */
    protected $user;
}

```

`AppBundle/Entity/User.php`

Add the file from the [FOSUserBundle Integration](fosuser-bundle.md#creating-a-user-entity-with-serialization-groups).

## Security Configuration

```yaml
# To get started with security, check out the documentation:
# http://symfony.com/doc/current/book/security.html
security:
    encoders:

>         FOS\UserBundle\Model\UserInterface: bcrypt

Reviewer: The `auto` mode should be used isn't it? https://symfony.com/blog/new-in-symfony-4-3-native-password-encoder

            id: fos_user.user_provider.username

    firewalls:
        # disables authentication for assets and the profiler, adapt it according to your needs
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false

        oauth_token:
            pattern: ^/oauth/v2/token
            security: false

        oauth_authorize:
            pattern: ^/oauth/v2/auth
            # Add your favorite authentication process here

        api_docs:
            pattern: ^/(|docs.json|docs.jsonld|index.json)$
            form_login: false
            provider: fos_user_bundle
            http_basic:
                realm: "Please enter your login data."

        api:
            pattern: ^/.+
            fos_oauth: true
            stateless: true
            anonymous: true

    access_control:
        - { path: ^/.+, roles: [IS_AUTHENTICATED_FULLY] }
```

## Routing Configuration

```yaml
api:
    resource: "."
    type: "api_platform"

fos_oauth_server_token:
    resource: "@FOSOAuthServerBundle/Resources/config/routing/token.xml"

fos_oauth_server_authorize:
    resource: "@FOSOAuthServerBundle/Resources/config/routing/authorize.xml"
```

> ## Create clients

Reviewer: ```suggestion
## Create Clients
```

> Add this class to `AppBundle/Command/CreateOAuthClientCommand.php` to generate oauth clients via console:

Reviewer: ```suggestion
Add this class to `api/src/Command/CreateOAuthClientCommand.php` to generate Oauth clients via console:
```

Other paths and namespaces must be updated accordingly.

```php
<?php

namespace AppBundle\Command;

use AppBundle\Entity\Client;
use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\HttpFoundation\Request;

> class CreateOAuthClientCommand extends ContainerAwareCommand

Reviewer 2: The bundle already provides [a similar command](https://github.com/FriendsOfSymfony/FOSOAuthServerBundle/blob/master/Command/CreateClientCommand.php) out of the box, isn't it?
